### PR TITLE
fix: 13929: Investigate ValidateLeafIndexHalfDiskHashMap failure

### PR DESCRIPTION
**Pull request by artemananiev**
Fix summary: see bug comments for details.

Testing: an existing unit test is modified to test a case, when both dirty and deleted leaf streams are empty.

Fixes: https://github.com/hashgraph/hedera-services/issues/13929
Signed-off-by: Artem Ananev <artem.ananev@swirldslabs.com>

### CODE REVIEW

File: platform-sdk/swirlds-jasperdb/src/main/java/com/swirlds/merkledb/MerkleDbDataSource.java [hunk 1]
```diff
@@ -1162,10 +1168,21 @@ private void writeLeavesToPathToKeyValue(
 
             // cache the record
             invalidateReadCache(leafRecord.getKey());
-        });
+        }
+        ;
```
Comment (OlegMazurov) [inline platform-sdk/swirlds-jasperdb/src/main/java/com/swirlds/merkledb/MerkleDbDataSource.java:1172]: Remove ;

File: platform-sdk/swirlds-jasperdb/src/main/java/com/swirlds/merkledb/MerkleDbDataSource.java [hunk 2]
```diff
@@ -1196,16 +1213,19 @@ private void writeLeavesToPathToKeyValue(
...
-        // end writing
-        final DataFileReader<VirtualLeafRecord<K, V>> pathToKeyValueReader = pathToKeyValue.endWriting();
-        statisticsUpdater.setFlushLeavesStoreFileSize(pathToKeyValueReader);
-        compactionCoordinator.compactPathToKeyValueAsync();
-        if (!isLongKeyMode) {
-            final DataFileReader<Bucket<K>> objectKeyToPathReader = objectKeyToPath.endWriting();
-            statisticsUpdater.setFlushLeafKeysStoreFileSize(objectKeyToPathReader);
-            compactionCoordinator.compactDiskStoreForObjectKeyToPathAsync();
+        }
+        ;
```
Comment (OlegMazurov) [inline platform-sdk/swirlds-jasperdb/src/main/java/com/swirlds/merkledb/MerkleDbDataSource.java:1217]: Ditto

File: platform-sdk/swirlds-jasperdb/src/main/java/com/swirlds/merkledb/MerkleDbDataSource.java [hunk 3]
```diff
@@ -1124,23 +1126,27 @@ private void writeHashes(final long maxValidPath, final Stream<VirtualHashRecord
...
-        }
+        // If both dirty/deleted leaves streams are empty, no new data files should be created
+        boolean startedWriting = false;
 
         // Iterate over leaf records
-        dirtyLeaves.sorted(Comparator.comparingLong(VirtualLeafRecord::getPath)).forEachOrdered(leafRecord -> {
+        final Iterator<VirtualLeafRecord<K, V>> dirtyIterator = dirtyLeaves
+                .sorted(Comparator.comparingLong(VirtualLeafRecord::getPath))
+                .iterator();
+        while (dirtyIterator.hasNext()) {
```
Comment (OlegMazurov) [inline platform-sdk/swirlds-jasperdb/src/main/java/com/swirlds/merkledb/MerkleDbDataSource.java:1140]: Now that you have iterators you can preserve the previous flow by checking `dirtyIterator.hasNext()` and `deletedIterator.hasNext()` and bailing out if both are false. Will avoid calling `startWriting()` on demand.

Comment (anthony-swirldslabs) [reply platform-sdk/swirlds-jasperdb/src/main/java/com/swirlds/merkledb/MerkleDbDataSource.java:1151]: +1. This way we don't need the `startedWriting` boolean flag here at all.

Comment (artemananiev) [reply platform-sdk/swirlds-jasperdb/src/main/java/com/swirlds/merkledb/MerkleDbDataSource.java:1151]: Done

**Review by anthony-swirldslabs — APPROVED** [review]

**Review by OlegMazurov — APPROVED** [review]

**Review by lpetrovic05 — APPROVED** [review]

**Pull request merged**